File: DLE.NET/TextureManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace DLE.NET
{
    //------------------------------------------------------------------------------
    //------------------------------------------------------------------------------
    //------------------------------------------------------------------------------

    public class TextureManager
    {
        public const int MAX_TEXTURES_D1 = 584;
        public const int MAX_TEXTURES_D2 = 910;

        public int MaxTextures { get { return DLE.IsD1File ? MAX_TEXTURES_D1 : MAX_TEXTURES_D2; } }
        public int Version { get { return DLE.IsD1File ? 0 : 1; } }
        public long Offset { get { return m_nOffsets [Version]; } }

        public Texture [] Textures { get { return m_textures [Version]; } }
        public PigTexture [] Infos { get { return m_info [Version]; } }
        public String [] Names { get { return m_names [Version]; } }

        public Texture [][] m_textures = new Texture [2][] { new Texture [MAX_TEXTURES_D1], new Texture [MAX_TEXTURES_D2] };
        public ushort [][] m_index = new ushort [2][];
        public String [][] m_names = new String [2][];
        public PigHeader [] m_header;
        public PigTexture [][] m_info = new PigTexture [2][] { null, null };
        public String [] m_pigFiles = new String [2] { "", "" };
        public uint[] m_nTextures = new uint [2] { 0, 0 };

        long [] m_nOffsets = new long [2] { 0, 0 };
        byte [] m_bmBuffer = new byte [512 * 512 * 4 * 32];
        List<Texture> m_extra = new List<Texture> ();
        String m_paletteName = "";

        //------------------------------------------------------------------------------

        public TextureManager ()
        {
            LoadIndex (0);
            LoadIndex (1);
        }

        //------------------------------------------------------------------------------

        public void Setup ()
        {
            Create (0);
            Create (1);
        }

        //------------------------------------------------------------------------------

        public void Create (int nVersion)
        {
            m_pigFiles [nVersion] = "";
            m_names [nVersion] = null;
            LoadNames (nVersion);
            m_header [nVersion] = new PigHeader (nVersion);
            LoadIndex (nVersion);
            m_textures [nVersion] = new Texture [MaxTextures];
        }

        //------------------------------------------------------------------------------

        public void Destroy (int nVersion)
        {
        }

        //------------------------------------------------------------------------------

        public Texture Texture (int nTexture)
        {
            return Textures [nTexture];
        }

        //------------------------------------------------------------------------------

        public String Name (int nTexture)
        {
            return Names [nTexture];
        }

        //------------------------------------------------------------------------------

        public PigTexture Info (int nTexture)
        {
            return Infos [nTexture];
        }

        //------------------------------------------------------------------------

        void Setup (int nVersion)
        {
            m_pigFiles [nVersion] = "";
            m_header [nVersion] = new PigHeader (nVersion);
            LoadNames (nVersion);
            LoadIndex (nVersion);
        }

        //------------------------------------------------------------------------------

        void LoadNames (int nVersion)
        {
            using (MemoryStream resource = new MemoryStream ((nVersion == 0) ? Properties.Resources.tnames : Properties.Resources.tnames2))
            {
                using (StreamReader reader = new StreamReader (resource))
                {
                    int j = MaxTextures;

                    m_names [nVersion] = new String [j];
                    for (int i = 0; i < j; i++)
                        m_names [nVersion] [i] = reader.ReadLine ();
                }
	        }
        }

        //------------------------------------------------------------------------------

        void LoadIndex (int nVersion)
        {
            using (MemoryStream resource = new MemoryStream ((nVersion == 0) ? Properties.Resources.texture : Properties.Resources.texture2))
            {
                using (BinaryReader reader = new BinaryReader (resource))
                {
                    m_nTextures [nVersion] = reader.ReadUInt32 ();
                    m_index [nVersion] = new ushort [m_nTextures [nVersion]];
                    for (int i = 0; i < m_index.Length; i++)
                        m_index [nVersion][i] = reader.ReadUInt16 ();
                }
            }
        }

        //------------------------------------------------------------------------------

        public PigTexture LoadInfo (FileStream fs, BinaryReader fp, short nTexture)
        {
            int nVersion = Version;
            if (m_info [nVersion] == null)
            {
                m_header [nVersion].Read (fp);
                m_info [nVersion] = new PigTexture [m_header [nVersion].nTextures];
                for (int i = 0; i < m_header [nVersion].nTextures; i++)
                    m_info [nVersion][i].Read (fp);
                m_nOffsets [nVersion] = fs.Position;
            }
            return m_info [nVersion][m_index [nVersion][nTexture] - 1];
        }

        //------------------------------------------------------------------------

        FileStream OpenPigFile ()
        {
            String filename  = DLE.descentPath [Version];
            if (!filename.Contains (".pig"))
                filename += "groupa.pig";
            FileStream fs = File.OpenRead (filename);
            using (BinaryReader fp = new BinaryReader (fs)) 
                {
                uint nOffset = fp.ReadUInt32 ();
                if (nOffset == 0x47495050) // 'PPIG' Descent 2 type
	                nOffset = 0;
                else if (nOffset < 0x10000)
	                nOffset = 0;
                fs.Seek (nOffset, SeekOrigin.Begin);
                }
            return fs;
        }

        //------------------------------------------------------------------------------

        void LoadTextures (int nVersion)
        {
        if (nVersion < 0) {
	        nVersion = Version;
	        m_pigFiles [nVersion] = DLE.descentPath [nVersion];
            m_paletteName = DLE.Palettes.Name;
	        m_info [nVersion] = null;
	        }
        Release (nVersion, true, false);
        FileStream fs = OpenPigFile ();
        for (short i = 0, j = (short) MaxTextures; i < j; i++)
	        m_textures [nVersion][i].Load (fs, i, nVersion);
        fs.Close ();
        }

        //------------------------------------------------------------------------------

        Texture AddExtra (ushort nIndex)
        {
	        Texture tex = new Texture();
            if (tex != null)
            {
                m_extra.Add (tex);
                tex.m_nIndex = nIndex;
            }
        return tex;
        }

        //------------------------------------------------------------------------

        void Release (int nVersion, bool bDeleteAll, bool bDeleteUnused)
        {
            // free any m_textures that have been buffered
            if (m_textures [nVersion] != null)
            {
                foreach (Texture tex in m_textures [nVersion])
                {
                    if (bDeleteUnused)
                    {
                        if (tex.m_bCustom && !tex.m_bUsed)
                            tex.Release ();
                    }
                    else
                    {
                        if (bDeleteAll || tex.m_bCustom)
                            tex.Release ();
                    }
                }
            }
        }

        //------------------------------------------------------------------------

        void Release (bool bDeleteAll, bool bDeleteUnused) 
        {
        // free any m_textures that have been buffered
        for (int i = 0; i < 2; i++) 
	        Release (i, bDeleteAll, bDeleteUnused);
        m_extra = new List<Texture> ();
        }


        //------------------------------------------------------------------------------

        void Blend (ref BGRA dest, BGRA src)
        {
        if (DLE.Palettes.SuperTransp (src))
	        dest.r =
	        dest.g =
	        dest.b =
	        dest.a = 0;
        else if (src.a > 0) {
	        if (src.a == 255)
		        dest = src;
	        else {
		        int b = 255 - src.a;
		        dest.r = (byte) (((int) dest.r * b + (int) src.r * src.a) / 255);
		        dest.g = (byte) (((int) dest.g * b + (int) src.g * src.a) / 255);
		        dest.b = (byte) (((int) dest.b * b + (int) src.b * src.a) / 255);
		        }
	        }
        }

        //-------------------------------------------------------------------------

        private struct tFrac
        {
            public int c, d;
        }

        int BlendTextures (short nBaseTex, short nOvlTex, Texture destTexP, int x0, int y0) 
        {

            short [] nTextures = new short [2];
            Texture [] texP = new Texture [2];
            BGRA [] bmBufP = destTexP.Buffer;
            BGRA [] srcDataP;
            int mode;
            int w, h, i, j, x, y, z, y1, s;
            tFrac scale, scale2;

            nTextures [0] = nBaseTex;
            nTextures [1] = (short) (nOvlTex & 0x3fff);
            mode = (int) (nOvlTex & 0xC000);

            for (i = 0; i < 2; i++) 
            {
                if ((nTextures [i] < 0) || (nTextures [i] >= MaxTextures))
                    nTextures [i] = 0;
                texP [i] = Texture (nTextures [i]);
            }
	
	        // Define bmBufP based on texture numbers and rotation
            destTexP.m_width = texP [0].m_width;
            destTexP.m_height = texP [0].m_height;
            destTexP.m_size = texP [0].m_size;
            destTexP.m_bValid = true;
            srcDataP = texP [0].Buffer;
            // if not rotated, then copy directly
            if (x0 == 0 && y0 == 0)
            {
                Buffer.BlockCopy (bmBufP, 0, srcDataP, 0, srcDataP.Length * 4);
            }
            else
            {
                // otherwise, copy bit by bit
                w = (int)texP [0].m_width;
                int l1 = y0 * w + x0;
                int l2 = (int)texP [0].m_size - l1;
                Buffer.BlockCopy (bmBufP, 0, srcDataP, l1, l2);
                Buffer.BlockCopy (bmBufP, l2, srcDataP, 0, l2);
                h = w;//texP [0].m_height;
                i = 0;
                for (y = 0; y < h; y++)
                    for (x = 0; x < w; x++)
                    {
                        z = (((y - y0 + h) % h) * w) + ((x - x0 + w) % w);
                        bmBufP [i] = srcDataP [z];
                    }
            }

            // Overlay texture 2 if present

            if (nTextures [1] == 0)
                return 0;
            srcDataP = texP [1].Buffer;
            if (texP [0].m_width == texP [1].m_width)
                scale.c = scale.d = 1;
            else if (texP [0].m_width < texP [1].m_width) 
            {
                scale.c = (int) (texP [1].m_width / texP [0].m_width);
                scale.d = 1;
            }
            else 
            {
                scale.d = (int) (texP [0].m_width / texP [1].m_width);
                scale.c = 1;
            }
            scale2.c = scale.c * scale.c;
            scale2.d = scale.d * scale.d;

            w = (int) texP [1].m_width / scale.c * scale.d;
            h = w;//texP [1].m_height / scale.c * scale.d;
            s = (int) (texP [1].m_width * texP [1].m_width)/*texP [1].m_size*/ / scale2.c * scale2.d;
            if ((x0 == 0) && (y0 == 0)) 
            {
                if (mode == 0) 
                {
                    i = 0;
                    for (y = 0; y < h; y++)
                    {
                        for (x = 0; x < w; x++, i++) 
                        {
                            bmBufP [i].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0x4000) 
                {
                    i = h - 1;
                    for (y = 0; y < h; y++, i--)
                    {
                        j = i;
                        for (x = 0; x < w; x++, j += w) 
                        {
                            bmBufP [j].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0x8000) 
                {
                    i = s - 1;
                    for (y = 0; y < h; y++)
                    {
                        for (x = 0; x < w; x++, i--) 
                        {
                            bmBufP [i].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0xC000) 
                {
                    i = (h - 1) * w;
                    for (y = 0; y < h; y++, i++)
                    {
                        j = i;
                        for (x = 0; x < w; x++, i -= w) 
                        {
                            bmBufP [j].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
            } 
            else 
            {
                if (mode == 0x0000) 
                {
                    for (y = 0; y < h; y++) 
                    {
                        y1 = ((y + y0) % h) * w;
                        for (x = 0; x < w; x++) 
                        {
                            bmBufP [y1 + (x + x0) % w].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0x4000) 
                {
                    for (y = h - 1; y >= 0; y--)
                    {
                        for (x = 0; x < w; x++) 
                        {
                            bmBufP [((x + y0) % h) * w + (y + x0) % w].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0x8000) 
                {
                    for (y = h - 1; y >= 0; y--) 
                    {
                        y1 = ((y + y0) % h) * w;
                        for (x = w - 1; x >= 0; x--) 
                        {
                            bmBufP [y1 + (x + x0) % w].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
                else if (mode == 0xC000) 
                {
                    for (y = 0; y < h; y++)
                    {
                        for (x = w - 1; x >= 0; x--) 
                        {
                            bmBufP [((x + y0) % h) * w + (y + x0) % w].Blend (srcDataP [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d]);
                        }
                    }
                }
            }
        return 0;
        }

        //------------------------------------------------------------------------

        bool HasCustom
        {
            get
            {
                Texture[] texs = Textures;
                for (int i = 0; i < MaxTextures; i++)
                    if (texs [i].m_bCustom)
                        return true;
                return false;
            }
        }

        //------------------------------------------------------------------------

        ushort CustomCount
        {
            get
            {
                Texture [] texs = Textures;
                ushort count = 0;
                for (int i = 0; i < MaxTextures; i++)
                    if (texs [i].m_bCustom)
                        count++;
                return count;
            }
        }

        //------------------------------------------------------------------------

        void MarkUsed ()
        {
	        int nVersion = DLE.IsD1File ? 0 : 1;
	        int i, j, h = MaxTextures;
            Segment [] segs = DLE.Mine.Segments;
            Texture [] texs = m_textures [Version];

            for (i = 0; i < h; i++)
	            texs [i].m_bUsed = false;

            for (i = 0; i < h; i++)
            {
                Side[] sides = segs [i].m_sides;
	            for (j = 0; j < 6; j++) 
                {
		            if ((sides [i].m_nChild < 0) || (sides [i].m_nWall != DLE.Mine.NO_WALL)) 
                    {
			            m_textures [nVersion][sides [i].m_nBaseTex & 0x1FFF].m_bUsed = true;
			            if ((sides [i].m_nOvlTex & 0x1FFF) != 0)
				            m_textures [nVersion][sides [i].m_nOvlTex & 0x1FFF].m_bUsed = true;
			        }
		        }
	        }

            Texture parentTex = null;

            for (i = 0; i < h; i++) 
            {
	            if (!texs [i].m_bFrame)
                    parentTex = texs [i];
                else if (texs [i].m_bCustom && !texs [i].m_bUsed)
                    texs [i].m_bUsed = (parentTex != null) && parentTex.m_bUsed;
	        }
        }

        //------------------------------------------------------------------------
        // remove unused custom m_textures

        void RemoveUnused ()
        {
	        int nCustom = CustomCount;

        if (nCustom > 0) {
	        MarkUsed ();
	        Release (Version, false, true);
	        int nRemoved = nCustom - CustomCount;
            String message = nRemoved + " custom textures " + ((nRemoved == 1) ? "was" : "were") + " removed";
            //INFOMSG (message);
            if (nRemoved > 0)
		        DLE.Modified = true;
	        }
        }

        //------------------------------------------------------------------------------

    }
}

File: texturemanager.cpp
#include <assert.h>

#include "stdafx.h"
#include "textures.h"
#include "cfile.h"
#include "dle-xp-res.h"
#include "global.h"
#include "palette.h"
#include "customtextures.h"
#include "texturemanager.h"
#include "dle-xp.h"

CTextureManager textureManager;

//------------------------------------------------------------------------
//------------------------------------------------------------------------
//------------------------------------------------------------------------

void CTextureManager::Setup (void)
{
names [0] = names [1] = null;
LoadTextures (0);
LoadTextures (1);
}

//------------------------------------------------------------------------

int CTextureManager::Version (void) 
{ 
return DLE.IsD1File () ? 0 : 1; 
}

//------------------------------------------------------------------------

void CTextureManager::Release (int nVersion, bool bDeleteAll, bool bDeleteUnused) 
{
// free any textures that have been buffered
CTexture* texP = &textures [nVersion][0];
for (int i = 0, h = MaxTextures (nVersion); i < h; i++, texP++) {
	if (bDeleteUnused) {
		if (texP->m_info.bCustom && !texP->m_info.bUsed)
			texP->Release ();
		}
	else {
		if (bDeleteAll || texP->m_info.bCustom)
			texP->Release ();
		}
	}		
}

//------------------------------------------------------------------------

void CTextureManager::Release (bool bDeleteAll, bool bDeleteUnused) 
{
// free any textures that have been buffered
for (int i = 0; i < 2; i++) 
	Release (i, bDeleteAll, bDeleteUnused);
for (CExtraTexture* p = extraTextures; p != null; ) {
	extraTextures = p->m_next;
	delete p;
	}
}

//------------------------------------------------------------------------

CFileManager* CTextureManager::OpenPigFile (int nVersion)
{
	CFileManager* fp = new CFileManager;
	char	filename [256];

strcpy_s (filename, sizeof (filename), nVersion ? descent2_path : descent_path);
if (!strstr (filename, ".pig"))
	strcat_s (filename, sizeof (filename), "groupa.pig");
if (fp->Open (filename, "rb")) {
	DEBUGMSG (" Reading texture: Texture file not found.");
	return null;
	}
uint nOffset = fp->ReadUInt32 ();
if (nOffset == 0x47495050) // 'PPIG' Descent 2 type
	nOffset = 0;
else if (nOffset < 0x10000)
	nOffset = 0;
fp->Seek (nOffset, SEEK_SET);
return fp;
}

//------------------------------------------------------------------------

int CTextureManager::MaxTextures (int nVersion)
{
return ((nVersion < 0) ? DLE.IsD2File () : nVersion) ? MAX_D2_TEXTURES : MAX_D1_TEXTURES;
}

//------------------------------------------------------------------------

bool CTextureManager::HasCustomTextures (void) 
{
CTexture* texP = &textures [DLE.FileType ()][0];

for (int i = MaxTextures (DLE.FileType ()); i; i--, texP++)
	if (texP->m_info.bCustom)
		return true;
return false;
}

//------------------------------------------------------------------------

int CTextureManager::CountCustomTextures (void) 
{
	int			count = 0;
	CTexture*	texP = &textures [DLE.FileType ()][0];

for (int i = MaxTextures (); i; i--, texP++)
	if (texP->m_info.bCustom)
		count++;
return count;
}

//------------------------------------------------------------------------------

bool CTextureManager::Check (int nTexture)
{
if ((nTexture >= 0) && (nTexture < MaxTextures ()))
	return true;
sprintf_s (message, sizeof (message), "Reading texture: Texture #%d out of range.", nTexture);
DEBUGMSG (message);
return false;
}

//------------------------------------------------------------------------------

void CTextureManager::Load (ushort nBaseTex, ushort nOvlTex)
{
if (Check (nBaseTex)) {
   textures [(int)DLE.FileType ()] [nBaseTex].Load (nBaseTex);
   if (Check (nOvlTex & 0x1FFF) && ((nOvlTex & 0x1FFF) != 0)) {
       Check ((ushort) (nOvlTex & 0x1FFF));
       textures [(int)DLE.FileType ()] [(ushort) (nOvlTex & 0x1FFF)].Load ((ushort) (nOvlTex & 0x1FFF));
		}
   }
}

//------------------------------------------------------------------------------

void CTextureManager::LoadNames (int nVersion)
{
	int nResource = (DLE.IsD1File ()) ? TEXTURE_STRING_TABLE_D1 : TEXTURE_STRING_TABLE_D2;
	CStringResource res;
	int j = MaxTextures (nVersion);

names [nVersion] = new char* [j];
for (int i = 0; i < j; i++) {
	res.Clear ();
	res.Load (nResource + i);
	int l = res.Length () + 1;
	names [nVersion][i] = new char [l];
	memcpy_s (names [nVersion][i], l, res.Value (), l);
	}
}

//------------------------------------------------------------------------------

int CTextureManager::LoadIndex (int nVersion)
{
	CResource res;	

ushort* indexP = (ushort *) res.Load (nVersion ? IDR_TEXTURE2_DAT : IDR_TEXTURE_DAT);
if (!indexP) {
	DEBUGMSG (" Reading texture: Could not load texture index.");
	return 1;
	}
// first long is number of textures
nTextures [nVersion] = *((uint*) indexP);
indexP += 2;
if (!(index [nVersion] = new ushort [nTextures [nVersion]])) {
	DEBUGMSG (" Reading texture: Could not allocate texture index.");
	return 3;
	}
for (uint i = 0; i < nTextures [nVersion]; i++)
	index [nVersion][i] = (*indexP++);
return 0;
}

//------------------------------------------------------------------------------

CPigTexture& CTextureManager::LoadInfo (CFileManager& fp, int nVersion, short nTexture)
{
if (info [nVersion] == null) {
	header [nVersion].Read (fp);
	info [nVersion] = new CPigTexture [header [nVersion].nTextures];
	for (int i = 0; i < header [nVersion].nTextures; i++)
		{
		info [nVersion][i].Read (fp, nVersion);
		nOffsets [nVersion] = fp.Tell ();
		}
	}
return info [nVersion][index [nVersion][nTexture] - 1];
}

//------------------------------------------------------------------------------

void CTextureManager::ReloadTextures (int nVersion)
{
if (nVersion < 0)
	Release (nVersion = Version (), false, false);

CFileManager* fp = OpenPigFile (nVersion);
for (int i = 0, j = MaxTextures (nVersion); i < j; i++)
	textures [nVersion][i].Load (i, nVersion, fp);
fp->Close ();
}

//------------------------------------------------------------------------------

void CTextureManager::LoadTextures (int nVersion)
{
header [nVersion] = CPigHeader (nVersion);
LoadNames (nVersion);
LoadIndex (nVersion);
#if USE_DYN_ARRAYS
textures [nVersion].Create (MaxTextures (nVersion));
#else
textures [nVersion] = new CTexture [MaxTextures (nVersion)];
#endif
ReloadTextures (nVersion);
}

//------------------------------------------------------------------------------
// textureManager.Define ()
//
// ACTION - Defines data with texture.  If bitmap_handle != 0,
//          then data is copied from the global data.  Otherwise,
//          the pig file defines bmBuf and then global memory
//          is allocated and the data is copied to the global memory.
//          The next time that texture is used, the handle will be defined.
//------------------------------------------------------------------------------

int CTextureManager::Define (short nBaseTex, short nOvlTex, CTexture *destTexP, int x0, int y0) 
{
	typedef struct tFrac {
		int	c, d;
	} tFrac;

	byte			*ptr;
	short			nTextures [2], mode, w, h;
	int			i, x, y, y1, offs, s;
	tFrac			scale, scale2;
	int			rc; // return code
	CTexture*	texP [2];
	byte			*bmBufP = destTexP->m_info.bmDataP;
	byte			c;
	int			fileType = DLE.FileType ();

nTextures [0] = nBaseTex;
nTextures [1] = nOvlTex & 0x3fff;
mode = nOvlTex & 0xC000;
for (i = 0; i < 2; i++) {
#if 0	
	ASSERT (textures [i] < MAX_TEXTURES);
#endif
	if ((nTextures [i] < 0) || (nTextures [i] >= MAX_TEXTURES))
		nTextures [i] = 0;
	// buffer textures if not already buffered
	texP [i] = &textures [fileType][nTextures [i]];
	if (!(texP [i]->m_info.bmDataP && texP [i]->m_info.bValid))
		if (rc = texP [i]->Load (nTextures [i]))
			return rc;
	}
	
	// Define bmBufP based on texture numbers and rotation
destTexP->m_info.width = texP [0]->m_info.width;
destTexP->m_info.height = texP [0]->m_info.height;
destTexP->m_info.size = texP [0]->m_info.size;
destTexP->m_info.bValid = 1;
ptr = texP [0]->m_info.bmDataP;
if (ptr) {
	// if not rotated, then copy directly
	if (x0 == 0 && y0 == 0) 
		memcpy (bmBufP, ptr, texP [0]->m_info.size);
	else {
		// otherwise, copy bit by bit
		w = texP [0]->m_info.width;
#if 1
		int	l1 = y0 * w + x0;
		int	l2 = texP [0]->m_info.size - l1;
		memcpy (bmBufP, ptr + l1, l2);
		memcpy (bmBufP + l2, ptr, l1);
#else
		byte		*dest = bmBufP;
		h = w;//texP [0]->m_info.height;
		for (y = 0; y < h; y++)
			for (x = 0; x < w; x++)
				*dest++ = ptr [(((y - y0 + h) % h) * w) + ((x - x0 + w) % w)];
#endif
		}
	}

// Overlay texture 2 if present

if (nTextures [1] == 0)
	return 0;
if (!(ptr = texP [1]->m_info.bmDataP))
	return 0;
if (texP [0]->m_info.width == texP [1]->m_info.width)
	scale.c = scale.d = 1;
else if (texP [0]->m_info.width < texP [1]->m_info.width) {
	scale.c = texP [1]->m_info.width / texP [0]->m_info.width;
	scale.d = 1;
	}
else if (texP [0]->m_info.width > texP [1]->m_info.width) {
	scale.d = texP [0]->m_info.width / texP [1]->m_info.width;
	scale.c = 1;
	}
scale2.c = scale.c * scale.c;
scale2.d = scale.d * scale.d;
offs = 0;
w = texP [1]->m_info.width / scale.c * scale.d;
h = w;//texP [1]->m_info.height / scale.c * scale.d;
s = (texP [1]->m_info.width * texP [1]->m_info.width)/*texP [1]->m_info.size*/ / scale2.c * scale2.d;
if (!(x0 || y0)) {
	byte *dest, *dest2;
	if (mode == 0x0000) {
		dest = bmBufP;
		for (y = 0; y < h; y++)
			for (x = 0; x < w; x++, dest++) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					*dest = c;
				}
		}
	else if (mode == (short) 0x4000) {
		dest = bmBufP + h - 1;
		for (y = 0; y < h; y++, dest--)
			for (x = 0, dest2 = dest; x < w; x++, dest2 += w) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					*dest2 = c;
				}
		}
	else if (mode == (short) 0x8000) {
		dest = bmBufP + s - 1;
		for (y = 0; y < h; y++)
			for (x = 0; x < w; x++, dest--) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					*dest = c;
				}
		}
	else if (mode == (short) 0xC000) {
		dest = bmBufP + (h - 1) * w;
		for (y = 0; y < h; y++, dest++)
			for (x = 0, dest2 = dest; x < w; x++, dest2 -= w) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					*dest2 = c;
				}
		}
	} 
else {
	if (mode == 0x0000) {
		for (y = 0; y < h; y++) {
			y1 = ((y + y0) % h) * w;
			for (x = 0; x < w; x++) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					bmBufP [y1 + (x + x0) % w] = c;
				}
			}
		}
	else if (mode == (short) 0x4000) {
		for (y = h - 1; y >= 0; y--)
			for (x = 0; x < w; x++) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					bmBufP [((x + y0) % h) * w + (y + x0) % w] = c;
				}
		}
	else if (mode == (short) 0x8000) {
		for (y = h - 1; y >= 0; y--) {
			y1 = ((y + y0) % h) * w;
			for (x = w - 1; x >= 0; x--) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					bmBufP [y1 + (x + x0) % w] = c;
				}
			}
		}
	else if (mode == (short) 0xC000) {
		for (y = 0; y < h; y++)
			for (x = w - 1; x >= 0; x--) {
				c = ptr [(y * scale.c / scale.d) * (w * scale.c / scale.d) + x * scale.c / scale.d];
				if (c != 255)
					bmBufP [((x + y0) % h) * w + (y + x0) % w] = c;
				}
			}
	}
return 0;
}

//------------------------------------------------------------------------

void CTextureManager::MarkUsedTextures (void)
{
	CSegment* segP = theMine->Segments (0);
	int nVersion = DLE.IsD1File () ? 0 : 1;
	int i, j, h = MaxTextures (nVersion);

for (i = 0; i < h; i++)
	textures [nVersion][i].m_info.bUsed = false;

for (i = theMine->SegCount (); i; i--, segP++) {
	CSide* sideP = segP->m_sides;
	for (j = 6; j; j--, sideP++) {
		if ((sideP->m_info.nChild < 0) || (sideP->m_info.nWall != NO_WALL)) {
			textures [nVersion][sideP->m_info.nBaseTex & 0x1FFF].m_info.bUsed = true;
			if ((sideP->m_info.nOvlTex & 0x1FFF) != 0)
				textures [nVersion][sideP->m_info.nOvlTex & 0x1FFF].m_info.bUsed = true;
			}
		}
	}

CTexture * texP, * parentTexP = null;

for (i = 0; i < h; i++) {
	texP = &textures [nVersion][i];
	if (!texP->m_info.bFrame)
		parentTexP = texP;
	else if (texP->m_info.bCustom && !texP->m_info.bUsed)
		texP->m_info.bUsed = (parentTexP != null) && parentTexP->m_info.bUsed;
	}
}

//------------------------------------------------------------------------
// remove unused custom textures

void CTextureManager::RemoveUnusedTextures (void)
{
	int nCustom = CountCustomTextures ();

if (nCustom) {
	MarkUsedTextures ();
	Release (false, true);
	int nRemoved = nCustom - CountCustomTextures ();
	sprintf_s (message, sizeof (message), "%d custom textures %s removed", nRemoved, (nRemoved == 1) ? "was" : "were");
	if (nRemoved)
		DLE.SetModified (TRUE);
	INFOMSG (message);
	}
}

//------------------------------------------------------------------------
//------------------------------------------------------------------------
//------------------------------------------------------------------------

//eof textures.cpp